### ISSUE TITLE
Ctrl-left mutates the command line

### ISSUE THREAD

**Issue by jstarks** [malxau/yori#46]
The Ctrl-Left functionality to move the cursor between words also expands environment variables and trims spaces. However, it doesn't actually re-render the command line immediately, so the mutations only show up after a subsequent editing command.

This can lead to some confusing and undesirable behavior, as seen in this video, where I use Ctrl-Left to jump past an environment variable and then press Del to delete the first character (which should be `%`):

![yori](https://user-images.githubusercontent.com/9548354/82525649-bbac2c80-9ae6-11ea-88a3-1793f4d9c945.gif)

**Comment by malxau** [comment]
I pushed commit 888f0bd3e1c52de024689a04d71f2eaf612b6b45 which fixes the redraw issue.  I take it from the title that you'd prefer that it didn't expand the environment variables, which will be a larger change.

**Comment by jstarks** [comment]
Thanks. Yes, expansion still seems undesirable to me, but at least with your fix it's easier to tell what's going on.

**Comment by malxau** [comment]
Okay, looks like I was wrong.  Eliminating the environment variable expansion here is commit 2c7ec92d999b6f69d48391adca4149f4f34405d8 and was quite straightforward.

**Comment by jstarks** [comment]
Cool, this is better, thanks! Ctrl-left will still collapse spaces, though--should that be made conditional too?

**Comment by malxau** [comment]
Note that in CMD, Ctrl-left moves a single word, whereas in Yori it moves a single argument.  This seemed more logical to me and is consistent with what the rest of the code is doing.  The way it's doing it is by parsing all of the arguments, finding which one the cursor should be, and reassembling the result - hence this behavior.  (Obviously my life would be easier if I just treated it as an opaque string and looked for spaces.)

But this is really just the tip of the iceberg.  The argument normalization you're seeing here also occurs when commands are executed, so those extra spaces are going to disappear later.  Note that Linux shells have to do this, because argc/argv are parsed by the parent process, but CMD didn't, because it's just passing the string to the child process to parse.  The rabbit hole I fell down that landed me here is trying to tab complete an argument while preserving later arguments, which requires understanding what an argument is, and reassembling multiple arguments, which results in duplicate space removal.  Having done that, I had an argc/argv parser, and used it everywhere.

Anyway, compare these commands:
```
echo foo          bar
echo "foo          bar"
```
On CMD, Yori, and bash.

Point being, I could make Ctrl-left/Ctrl-right logic all about context-blind spaces like CMD, but if you really wanted to keep duplicate spaces in the command to execute it'd take far more than the code here.  So far I haven't regretted the choice I made with argument parsing - it's a huge point of difference between Yori and CMD but I haven't seen anything break, presumably because I did what Linux does, so programs supporting it can't depend on multiple spaces.

Supporting the CMD way requires a parser that can find the "first" argument (command to execute), then assemble text following into a blob, after carefully fishing out redirects, and the blob is terminated by a pipe or command separator (&&, || type thing.)  That parser would presumably exist alongside the current one, which is still needed for tab completing command arguments.

(Trivia question: how many spaces in the output file when you execute this command on the three shells instead)...
```
echo foo  >  file  bar
```

**Comment by jstarks** [comment]
Ha, yes. I feel like bash and Yori (and PowerShell) do the right thing, or at least what I'm used to. I thing DOS/Windows got it wrong by leaving (most) argument parsing up to the app (related problem: Yori builtins appear to support ~ meaning %USERPROFILE%, but the expansion is not done by the shell itself so I can't actually rely on it).

I think notably because PowerShell gets away with it, probably Yori can too.

But regardless of the string that is passed to CreateProcess, a more clever parsing algorithm could preserve the spaces during tab completion and cursor navigation.

**Comment by malxau** [comment]
The original issue here has been fixed in the 1.30 release.  Please reactivate this issue or file a new issue if there are further issues related to it.
